[Dynamo] a dyanmo backend based on ONNXRuntime

wschin opened the pull request:
This PR migrates the dynamo backend developed under ONNXRuntime into PyTorch. The ultimate goal is to replace legacy `onnxrt` in dynamo with dynamo compiler from ONNXRuntime team.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

wschin committed:
Replace onnxrt with DORT

wschin committed:
Clean code and format

Follow main

wschin committed:
Move files to ONNX folder

wschin committed:
Supress lint warnings

make _backend a module

wschin committed:
Add tests

wschin committed:
Test shared and local backends

wschin committed:
Change TODO name

thiagocrepaldi commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+    # so we still need to check if it exists.
+    importlib.import_module("onnxscript")
+
+    _NP_DTYPE = {
+        torch.float16: np.float16,
+        torch.float32: np.float32,
+        torch.float64: np.float64,
+        torch.uint8: np.uint8,
+        torch.int8: np.int8,
+        torch.int16: np.int16,
+        torch.int32: np.int32,
+        torch.int64: np.longlong,
+        torch.bool: np.bool_,
+    }
+    _ONNX_ELEMENT_TYPE_TO_TORCH_DTYPE = {
+        onnx.TensorProto.FLOAT: torch.float32,  # type: ignore[attr-defined]
+        onnx.TensorProto.FLOAT16: torch.float16,  # type: ignore[attr-defined]
+        onnx.TensorProto.DOUBLE: torch.float64,  # type: ignore[attr-defined]
+        onnx.TensorProto.BOOL: torch.bool,  # type: ignore[attr-defined]
+        onnx.TensorProto.UINT8: torch.uint8,  # type: ignore[attr-defined]
+        onnx.TensorProto.INT8: torch.int8,  # type: ignore[attr-defined]
+        onnx.TensorProto.INT16: torch.int16,  # type: ignore[attr-defined]
+        onnx.TensorProto.INT32: torch.int32,  # type: ignore[attr-defined]
+        onnx.TensorProto.INT64: torch.int64,  # type: ignore[attr-defined]
+    }
+    _TORCH_DTYPE_TO_ONNX_ELEMENT_TYPE = {
+        value: key for key, value in _ONNX_ELEMENT_TYPE_TO_TORCH_DTYPE.items()
+    }
```
Can this be consolidated/reused at https://github.com/pytorch/pytorch/blob/main/torch/onnx/_internal/fx/type_utils.py?

wschin replied:
Yep, I will move them to `type_utils.py`.

thiagocrepaldi commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+
+
+def has_onnxruntime():
+    return _SUPPORT_ONNXRT
```
should this also `import onnxruntime` in addition to `import onnxscript`?

wschin replied:
Yes, it's checked.
```
try:
    # Use try-except to initialize package-dependent global variables.
    from typing import Any, Dict, List, Mapping, Optional, Set, Tuple, Union

    import onnx
    import onnxruntime  # type: ignore[import]
    from onnxruntime.capi import _pybind_state as ORTC  # type: ignore[import]

    # This is not use directly in DORT but needed by underlying exporter,
    # so we still need to check if it exists.
    importlib.import_module("onnxscript")
```

thiagocrepaldi commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+
+
+class OrtOperatorSupport(OperatorSupport):
+    """
+    Operator support for ONNXRuntime backend. It has two-level of support decision.
+    One is via support_dict and the other one is via extra_support_dict. The logic
+    of using support_dict is implemented in OrtOperatorSupport and extra_support_dict
+    is used by OperatorSupport.is_node_supported.
```
```suggestion
    """Operator support for ONNXRuntime backend.

    It has two-level of support. One is via `support_dict` and the other one is via `extra_support_dict`. The logic
    of using `support_dict` is implemented in `OrtOperatorSupport` and `extra_support_dict`
    is used by `OperatorSupport.is_node_supported`.
```
nit

thiagocrepaldi commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+                type(node.target),
+            )
+            return True
+        logger.info(
```
```suggestion
        logger.warn(
```
maybe a warn?

wschin replied:
Ok.

thiagocrepaldi commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+                type(node.target),
+            )
+            return True
+        logger.info(
```
```suggestion
        logger.warn(
```

thiagocrepaldi replied:
out of curiosity. what is the consequence of not supporting a node? would it cause a graph break?

wschin replied:
Yep. Graph break.

thiagocrepaldi commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+        if ep == "CPUExecutionProvider":
+            # Lowest priority.
+            return 2
+        if ep == "CUDAExecutionProvider":
```
for my knowledge, is rocm supported?

wschin replied:
Yes but never tested. To trigger ROCM, user has to specified `OrtBackend.ep = "ROCMExecutionProvider`.

thiagocrepaldi commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+        return pth_outputs
+
+
+def _assert_allclose_with_detailed_error_message(
```
Can Pytorch's [assertEqual](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_utils.py#L3188) be used instead? maybe [torch.testing.assert_close](https://github.com/pytorch/pytorch/blob/main/torch/testing/_comparison.py#L1250)?

wschin replied:
Sure. Will use torch.testing.assert_close.

wschin committed:
Capture exception when missing onnxscript

wschin committed:
Move onnxscript-dependent packages into try-except

abock commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+
+    Example usages:
+        import torch
+        from onnxruntime.training.torchdynamo.register_backend import make_aot_ort
```
Example needs updating to reflect this PR. Would be good if the examples can be shown using the torch.compile decorator?

BowenBao replied:
+1 plz use xdoctest style such that example code is checked by CI. https://github.com/pytorch/pytorch/blob/3a07dfde488ba974dc7ed7f8f24e9bcf64ada053/torch/onnx/_internal/exporter.py#L413-L437

wschin replied:
Here is the new example.
```
         # xdoctest: +REQUIRES(env:TORCH_DOCTEST_ONNX) 
         >>> import copy
         >>> import torch 
         >>> from torch.onnx._backend.core import make_aot_ort
         >>> class MyModel(torch.nn.Module):  # Dummy model 
         ... def __init__(self) -> None: 
         ...     super().__init__() 
         ...     self.linear = torch.nn.Linear(2, 2) 
         ... def forward(self, x): 
         ...     out = self.linear(x) 
         ...     return out 
         >>> model = MyModel()
         >>> aot_ort, _ = make_aot_ort(dynamic = True)
         >>> compiled_model = torch.compile(copy.deepcopy(model), backend=aot_ort, dynamic=True)
         >>> x = torch.randn(2, 2, 2)
         >>> y_baseline = model(x)
         >>> y_compiled = compiled_model(x)
         >>> torch.testing.assert_close(y_baseline, y_compiled)
```

BowenBao commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+    return (
+        aot_autograd(
+            fw_compiler=ort_backend,
+            partition_fn=min_cut_rematerialization_partition,
```
what does this partition do and why we need it?

wschin replied:
Explained in comment.

BowenBao commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+            else:
+                output_devices = _get_onnx_devices((prim_outputs,))
+
+            execution_info_per_session = OrtExecutionInfoPerSession(
```
TODO: Looks like a good blob of info to create diagnostic and log.

titaiwangms commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+        #  2. call exporter's API to convert `torch.fx.GraphModule` to ONNX model
+        #     (see onnxfunction_dispatcher passed to FxOnnxInterpreter.run below).
+        if onnx_exporter_options is None:
+            onnx_exporter_options = torch.onnx._internal.exporter.ExportOptions()
```
Sorry, nvm

```suggestion
            onnx_exporter_options = torch.onnx.ExportOptions()
```

wschin committed:
Address thiago's comments

wschin committed:
Split the introduction of DORT and deprecation of onnxrt

wschin commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+        return self.compile(graph_module, args)
+
+
+def make_aot_ort(dynamic: bool = True):
```
No. In `aot_autograd`'s signature, we see no type annotations. However, intellisense is able to infer it from the code. As this function calls `aot_autograd`, we cannot use type annotation so that the inferred types can be propagated by intellisense.

![image](https://github.com/pytorch/pytorch/assets/3524474/ab557bc1-c29f-48dc-a397-f8058e9ce487)

wschin commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+            return onnx_outputs[0]
+
+    def compile(self, graph_module: torch.fx.GraphModule, args) -> torch.fx.GraphModule:
+        # FX graph based partitioning based on ONNX supported ops.
```
Sure.
```
        # FX graph based partitioning based on ONNX supported ops.
        # Given a graph module
        #  GraphModule0
        #   node_0
        #   node_1
        #   node_2
        #   node_3
        #   node_4
        # If only node_2 is not supported by ONNX, this graph module will be partitioned into
        #  GraphModule0
        #   GraphModule1
        #    node_0
        #    node_1
        #   node_2
        #   GraphModule2
        #    node_3
        #    node_4
        # by calling CapabilityBasedPartitioner.partition_and_fuse.
        # Then, GraphModule1's and GraphModule2's forward method (GraphModule._wrapped_call)
        # will be replaced by OrtBackend._ort_accelerated_call to delegate computation to ORT.
```

wschin commented on torch/_dynamo/backends/onnxrt.py:
```diff
@@ -1,131 +1,6 @@
-    return _call
+if has_onnxruntime():
+    aot_ort, ort = make_aot_ort(dynamic=True)
+    register_backend(name="onnxrt", compiler_fn=aot_ort)
```
Nop. This PR now doesn't register `OrtBackend` as `onnxrt` in dynamo (so using `backend="onnxrt"` will lead to the old backend). In the next PR, we will do the registration and add document.

wschin commented on torch/_dynamo/backends/onnxrt.py:
```diff
@@ -1,131 +1,6 @@
-
-    return _call
+if has_onnxruntime():
+    aot_ort, ort = make_aot_ort(dynamic=True)
```
Good suggestion, but you need to create `OrtBackend` before calling `torch.compile`. As `OrtBackend` has been created, you can't set its `dynamic` when calling `torch.compile`.

wschin commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+
+    def __init__(
+        self,
+        ep: str = "CPUExecutionProvider",
```
# Default execution provider such as "CUDAExecutionProvider" or "CPUExecutionProvider".
        # It's used to create a new inference session when we can't find any device information
        # from arguments or the graph module captured by dynamo.
        self.ep = ep

wschin commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+        # should be supported, which is currently done through dlpack. Note that dlpack might not support a custom torch device.
+        # It can be avoided by allowing for preallocation for output buffers allocated by a custom aten allocator,
+        # and use the preallocated output buffers for InferenceSession not holding any ownership for them.
+        self.preallocate_output = preallocate_output
```
OK.

wschin commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+        onnx_exporter_options=ExportOptions(dynamic_shapes=dynamic)
+    )
+    return (
+        aot_autograd(
```
It's hard to tell... aot_autograd itself is not documented and can be changed but I will try putting my understanding here.

wschin committed:
Address bowen's comments

wschin commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+                _assert_allclose_with_detailed_error_message(
+                    onnx_outputs[0], baseline_outputs
+                )
+            return onnx_outputs[0]
```
I have no plan to avoid repeated codes since it will dramatically reduce readability for me. You can see for corresponding function calls in this if-else block, they are different at least at one argument. If you want to merge them, you will need a flag like
```
        if isinstance(prim_outputs, tuple) and all(isinstance(elem, torch.Tensor) for elem in prim_outputs):
            original_graph_module_output_is_a_tuple_of_tensors = True
        elif isinstance(prim_outputs, torch.Tensor):
            original_graph_module_output_is_a_tuple_of_tensors = False
        else:
            raise RuntimeError(
                f"prim_outputs must be either a tuple of tensors or a tensor. prim_outputs={prim_outputs}"
            )
```

and use it to create an if-else for each function call (_run_onnx_session_with_ortvaluevector, torch.testing.assert_close). It will end up with independent 3 if-else blocks, which read very bad. You probably can hide some logics inside `_run_onnx_session_with_ortvaluevector` but I only hide when there is no way to avoid.

BowenBao replied:
Below is easier to maintain and I don't think it is "dramatically reducing readability".
```python
# ORT always returns a tuple of outputs. If the original output is a tensor,
# ORT output's first element must be extracted and returned. Otherwise, type
# mismatch may happen in downstream computation.
is_single_tensor_output = isinstance(prim_outputs, torch.Tensor)
normalized_prim_outputs = (prim_outputs,) if is_single_tensor_output else prim_outputs
assert isinstance(normalized_prim_outputs, tuple)
assert all(isinstance(elem, torch.Tensor) for elem in normalized_prim_outputs)

_nvtx_range_push("run_onnx_session_with_ortvaluevector")
onnx_outputs = _run_onnx_session_with_ortvaluevector(
    onnx_session,
    input_names,
    args,
    input_devices,
    output_names,
    normalized_prim_outputs,
    output_devices,
    self.preallocate_output,
)
_nvtx_range_pop()
if self._assert_allclose_to_baseline:
    # Compute baseline.
    baseline_outputs = torch._prims.executor.execute(
        graph_module, *args, executor="aten"
    )
    normalized_baseline_ouptuts = (baseline_outputs,) if is_single_tensor_output else baseline_outputs
    # Ensure every output tensor is close to the corresponding baseline.
    for onnx_output, baseline_output in zip(onnx_outputs, normalized_baseline_ouptuts):
        torch.testing.assert_close(onnx_output, baseline_output)
return onnx_outputs[0] if is_single_tensor_output else onnx_outputs
```

wschin replied:
Done.

wschin commented on test/onnx/test_dynamo_with_onnxruntime_backend.py:
```diff
@@ -0,0 +1,139 @@
+# Owner(s): ["module: onnx"]
```
Ok.

wschin commented on torch/onnx/_backend/core.py:
```diff
@@ -0,0 +1,852 @@
+        # create_onnx_friendly_decomposition_table(...) in
+        # torch/onnx/_internal/fx/decomposition_table.py.
+        support_dict = torch.onnx._internal.fx.decomposition_table._create_onnx_supports_op_overload_table(
+            # This is identical to self.resolved_onnx_exporter_options.onnxfunction_dispatcher.onnx_registry.
```
removed.

wschin committed:
Address ti-tai's comments

wschin committed:
defer type annotation to runtime

defer accessing onnxruntime in type annotation

Add doc string to compile

More doc strings

wschin committed:
Make onnx_test_common importable from sub-folders

wschin committed:
Merge duplicated code and add test for its extra safety

Fix a test

wschin committed:
Use training's package since push_batch_pack is not in onnxruntime.
Once onnxruntime PR 17058 is merged, this change can be reverted.

wschin committed:
Merge branch 'main' into onnxrt-reborn

wschin committed:
Avoid onnx-weekly being overrided by onnx from installing onnxruntime

wschin committed:
Merge branch 'onnxrt-reborn' of https://github.com/wschin/pytorch into onnxrt-reborn

wschin committed:
Fix CI package install line

wschin committed:
Fix test

wschin committed:
skip ort bug

wschin committed:
Disable graph fusion since ORT has bug

wschin committed:
Use inference-only session.run if onnxruntime is installed

abock reviewed (changes requested):
In addition to @BowenBao's comments, I'd like to see the `_backend` folder renamed to `_internal/onnxruntime`. In the context of the rest of `torch.onnx` the name `_backend` doesn't tell me anything.

wschin committed:
Move and rename core.py to onxruntime.py

wschin committed:
Create issues for TODs

wschin commented:
@abock, done. TODO has issue now.

abock commented on torch/onnx/_internal/onnxruntime.py:
```diff
@@ -0,0 +1,881 @@
+         # xdoctest: +REQUIRES(env:TORCH_DOCTEST_ONNX)
+         >>> import copy
+         >>> import torch
+         >>> from torch.onnx._backend.core import make_aot_ort
```
This is now out of date due to the rename 😆 

```python
from torch.onnx._internal.onnxruntime import make_aot_ort
```

wschin replied:
Fixed and run well locally.

thiagocrepaldi commented on test/onnx/dynamo/test_dynamo_with_onnxruntime_backend.py:
```diff
@@ -0,0 +1,201 @@
+        torch._dynamo.reset()
+
+    def tearDown(self):
+        torch._dynamo.reset()
+        super().tearDown()
```
```suggestion
        super().tearDown()
        torch._dynamo.reset()
```

thiagocrepaldi commented on test/onnx/dynamo/test_dynamo_with_onnxruntime_backend.py:
```diff
@@ -0,0 +1,201 @@
+                for baseline_elem, result_elem in zip(baseline_result, result):
+                    torch.testing.assert_close(baseline_elem, result_elem)
+
+    def _check_counting_information(
```
```suggestion
    def _assert_counting_information(
```

thiagocrepaldi commented on torch/onnx/_internal/onnxruntime.py:
```diff
@@ -0,0 +1,881 @@
+
+    # This is not use directly in DORT but needed by underlying exporter,
+    # so we still need to check if it exists.
+    importlib.import_module("onnxscript")
```
You could use [torch.onnx._internal.exporter._assert_dependencies](https://github.com/pytorch/pytorch/blob/main/torch/onnx/_internal/exporter.py#L916) to check that

wschin replied:
I feel better no to touch anything under `torch.onnx` in this block.

thiagocrepaldi commented on torch/onnx/_internal/onnxruntime.py:
```diff
@@ -0,0 +1,881 @@
+        return False
+
+
+def _move_placeholder_to_front(graph_module: torch.fx.GraphModule) -> None:
```
Can [MovePlaceholderToFront](https://github.com/pytorch/pytorch/blob/main/torch/onnx/_internal/fx/passes/virtualization.py#L12) be reused?

wschin replied:
yep. replaced.

thiagocrepaldi commented on torch/onnx/_internal/onnxruntime.py:
```diff
@@ -0,0 +1,881 @@
+
+
+def _replace_to_copy_with_to(fx_module: torch.fx.GraphModule) -> None:
+    # aten._to_copy doesn't have exporter so we replace it with aten.to.
```
Maybe a TODO so that torchlib can (maybe) handle it?

wschin replied:
Tracked with issue when it's called.

thiagocrepaldi reviewed (approved):
🚢

wschin committed:
Merge remote-tracking branch 'origin/main' into onnxrt-reborn

wschin committed:
Address comments

thiagocrepaldi commented:
@pytorchbot merge
